compiler/src/java_plugin/cpp/java_generator.cpp

```diff
@@ -534,72 +561,74 @@ static void PrintStub(
     Printer* p, StubType type) {
   const std::string service_name = service->name();
   (*vars)["service_name"] = service_name;
-  (*vars)["abstract_name"] = service_name + "ImplBase";
   std::string stub_name = service_name;
-  std::string client_name = service_name;
   std::string stub_base_class_name = "AbstractStub";
+  std::string interface_name = service_name;
   CallType call_type;
   bool impl_base = false;
   bool interface = false;
   switch (type) {
+    case ASYNC_INTERFACE:
+      call_type = ASYNC_CALL;
+      interface = true;
+      impl_base = true;
+      stub_name += "Async";
+      break;
     case ABSTRACT_CLASS:
       call_type = ASYNC_CALL;
       impl_base = true;
+      interface_name += "Async";
+      (*vars)["abstract_name"] = service_name + "ImplBase";
       break;
     case ASYNC_CLIENT_IMPL:
       call_type = ASYNC_CALL;
       stub_name += "Stub";
       stub_base_class_name = "AbstractAsyncStub";
+      interface_name += "Async";
       break;
-    case BLOCKING_CLIENT_INTERFACE:
-      interface = true;
-      FALLTHROUGH;
     case BLOCKING_CLIENT_IMPL:
       call_type = BLOCKING_CALL;
       stub_name += "BlockingStub";
-      client_name += "BlockingClient";
       stub_base_class_name = "AbstractBlockingStub";
+      interface_name += "Blocking";
       break;
-    case FUTURE_CLIENT_INTERFACE:
-      interface = true;
-      FALLTHROUGH;
     case FUTURE_CLIENT_IMPL:
       call_type = FUTURE_CALL;
       stub_name += "FutureStub";
-      client_name += "FutureClient";
       stub_base_class_name = "AbstractFutureStub";
+      interface_name += "Future";
       break;
-    case ASYNC_INTERFACE:
-      call_type = ASYNC_CALL;
-      interface = true;
-      stub_name += "Stub";
-      stub_base_class_name = "AbstractAsyncStub";
+    case BLOCKING_CLIENT_INTERFACE:
+    case FUTURE_CLIENT_INTERFACE:
+      GRPC_CODEGEN_FAIL << "Intentionally not creating StubType: " << type;
       break;
     default:
       GRPC_CODEGEN_FAIL << "Cannot determine class name for StubType: " << type;
   }
   (*vars)["stub_name"] = stub_name;
-  (*vars)["client_name"] = client_name;
   (*vars)["stub_base_class_name"] = (*vars)[stub_base_class_name];
+  (*vars)["interface_name"] = interface_name;
 
   // Class head
-  if (!interface) {
-    GrpcWriteServiceDocComment(p, service);
-  }
+    GrpcWriteServiceDocComment(p, service, type);
 
   if (service->options().deprecated()) {
     p->Print(*vars, "@$Deprecated$\n");
   }
 
-  if (impl_base) {
+  if (interface) {
     p->Print(
         *vars,
-        "public static abstract class $abstract_name$"
-        " implements $BindableService$ {\n");
+        "public interface $stub_name$ {\n");
+  } else if (impl_base) {
+    p->Print(
+        *vars,
+        "public static abstract class $abstract_name$\n"
+        " implements $BindableService$, $interface_name$ {\n");
```

Comment:
The single-space was not indentation previously; it was just the separator after the class name. Either the `\n` on the previous line should be removed or we add more indentation here. Ditto below.